sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala

```diff
@@ -64,33 +65,143 @@ class UnivocityGenerator(
+          }
+          var i = 1
+          while (i < array.numElements()) {
+            builder.append(", ")
+            if (array.isNullAt(i)) {
+              if (nullAsQuotedEmptyString) {
+                builder.append(options.nullValue)
+              } else {
+                builder.append(null.asInstanceOf[String])
+              }
+            } else {
+              builder.append(elementConverter(array, i))
+            }
+            i += 1
+          }
+        }
+        builder.append("]")
+        builder.toString()
+
+    case MapType(kt, vt, _) =>
+      val keyConverter = makeConverter(kt)
+      val valueConverter = makeConverter(vt)
+      (getter, ordinal) =>
+        val map = getter.getMap(ordinal)
+        val builder = new StringBuilder
+        builder.append("{")
+        if (map.numElements() > 0) {
+          val keyArray = map.keyArray()
+          val valueArray = map.valueArray()
+          builder.append(keyConverter(keyArray, 0))
+          builder.append(" -> ")
+          if (valueArray.isNullAt(0)) {
+            if (nullAsQuotedEmptyString) {
+              builder.append(options.nullValue)
+            } else {
+              builder.append(null.asInstanceOf[String])
+            }
+          } else {
+            builder.append(valueConverter(valueArray, 0))
+          }
+          var i = 1
+          while (i < map.numElements()) {
+            builder.append(", ")
+            builder.append(keyConverter(keyArray, i))
+            builder.append(" -> ")
+            if (valueArray.isNullAt(i)) {
+              if (nullAsQuotedEmptyString) {
+                builder.append(options.nullValue)
+              } else {
+                builder.append(null.asInstanceOf[String])
+              }
+            } else {
+              builder.append(valueConverter(valueArray, i))
+            }
+            i += 1
+          }
+        }
+        builder.append("}")
+        builder.toString()
+
+    case StructType(fields) =>
+      val converters = fields.map(_.dataType).map(makeConverter)
+      (getter, ordinal) =>
+        val row = getter.getStruct(ordinal, fields.length)
+        val builder = new StringBuilder
+        builder.append("{")
+        if (row.numFields > 0) {
+          if (row.isNullAt(0)) {
+            if (nullAsQuotedEmptyString) {
+              builder.append(options.nullValue)
+            } else {
+              builder.append(null.asInstanceOf[String])
+            }
+          } else {
+            builder.append(converters(0)(row, 0))
+          }
+          var i = 1
+          while (i < row.numFields) {
+            builder.append(", ")
+            if (row.isNullAt(i)) {
+              if (nullAsQuotedEmptyString) {
+                builder.append(options.nullValue)
+              } else {
+                builder.append(null.asInstanceOf[String])
+              }
+            } else {
+              builder.append(converters(i)(row, i))
```

Comment:
Can we add a function for the common null handling code?

Reply:
Okay, I found a bug in the original (Before applying `this PR`) `to_csv`, as follows:
<img width="985" alt="image" src="https://github.com/apache/spark/assets/15246973/7491d868-3236-4518-8b27-7aed769d26e4">
https://github.com/apache/spark/blob/227a50a1766ac1476b0031e1c60d2604eccdb9a7/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala#L100-L104
<img width="538" alt="image" src="https://github.com/apache/spark/assets/15246973/e576e7e0-a03b-497f-929a-cf8b1cdcfcc8">

when `row.isNullAt(i)` && `nullAsQuotedEmptyString == false`, 
`values(i)` is not assigned, so it is the default value null, which seems incorrect?

<img width="859" alt="image" src="https://github.com/apache/spark/assets/15246973/68d263e9-934d-436f-b79d-1a04b6cd5b76">
According to this interpretation, we should assign value `""`.

@cloud-fan Should we fix it in this PR or open a separate PR to handle it?

Reply:
Please let me check a possible issue before merging.